### ISSUE TITLE
[BUG] Miscompilation of code involving references and exceptions

### ISSUE THREAD

**Issue by vouillon** [ocsigen/js_of_ocaml#1354]
I have found out what the block handler was for... This following piece of code is miscompiled. The issue is that the OCaml bytecode program does not allocate a reference but update the stack in place. But we no longer keep track of this update.
```ocaml
let x = ref 0 in
try
  incr x;
  raise Exit
with Exit ->
  Format.eprintf "%d@." !x
```
You have to compile it without the `-g` option. For instance:
```shell
ocamlc bug.ml -o bug.byte && js_of_ocaml bug.byte && ./bug.byte && node bug.js
```
The above command outputs `1` (correct) then `0` (wrong):

**Comment by vouillon** [comment]
I think that this piece of code, with nested exception handlers, has always been miscompiled.
```ocaml
let x = ref 0 in
try
  (try incr x ; raise Exit with _ -> ());
  raise Exit
with Exit ->
  Format.eprintf "%d@." !x
```

**Comment by hhugo** [comment]
I wonder if we could make that logic more explicit by keeping track of `ASSIGN`ed variable that have been previously captured in an exception handler. Maybe using a new instruction 
```ocaml
type instr =
  | Let of Var.t * expr
  | Set_field of Var.t * int * Var.t
  | Offset_ref of Var.t * int
  | Array_set of Var.t * Var.t * Var.t
  | Update_pushtrap_cont  of Var.t * Var.t
```

**Comment by hhugo** [comment]
The following program was never compiled correctly.
```ocaml
let x = ref 0 in
try
  incr x ;
  (try incr x ; raise Exit with _ -> Format.eprintf "%d@." !x);
  raise Exit
with Exit ->
  Format.eprintf "%d@." !x
```

I have wip fix in #1355
